src/OpenTelemetry.Exporter.Geneva/Internal/TableNameSerializer.cs

```diff
@@ -0,0 +1,313 @@
+        MessagePackSerializer.WriteStr8Header(buffer, 0, validNameLength);
+
+        return cursor;
+    }
+
+    private byte[] ResolveTableMappingForCategoryName(string categoryName)
+    {
+        var tableNameCache = this.m_tableNameCache;
+
+        if (tableNameCache.TryGetValue(categoryName, out byte[] tableName))
+        {
+            return tableName;
+        }
+
+        return this.ResolveTableMappingForCategoryNameRare(categoryName);
+    }
+
+    private byte[] ResolveTableMappingForCategoryNameRare(string categoryName)
+    {
+        lock (this.m_lockObject)
+        {
+            var tableNameCache = this.m_tableNameCache;
+
+            // Check if another thread added the mapping while we waited on the
+            // lock.
+            if (tableNameCache.TryGetValue(categoryName, out byte[] tableName))
+            {
+                return tableName;
+            }
+
+            byte[] mappedTableName = null;
+
+            // If user configured table name mappings run resolution logic.
+            if (this.m_tableMappings != null
+                && !this.m_tableMappings.TryGetValue(categoryName, out mappedTableName))
+            {
+                // Find best match if an exact match was not found.
+
+                string currentKey = null;
+
+                foreach (var mapping in this.m_tableMappings)
+                {
+                    if (!categoryName.StartsWith(mapping.Key, StringComparison.OrdinalIgnoreCase))
+                    {
+                        continue;
+                    }
+
+                    if (currentKey == null || mapping.Key.Length >= currentKey.Length)
+                    {
+                        currentKey = mapping.Key;
+                        mappedTableName = mapping.Value;
+                    }
+                }
+            }
+
+            mappedTableName ??= !this.m_shouldPassThruTableMappings
+                ? this.m_defaultTableName
+                : s_passthroughTableName;
+
+            if (mappedTableName.Length <= 0 && tableNameCache.CachedSanitizedTableNameCount < MaxCachedSanitizedTableNames)
+            {
+                // We resolved to a wildcard which is pass-through mode.
+
+                Span<byte> stackBuffer = stackalloc byte[64];
+
+                int bytesWritten = SanitizedCategoryName(stackBuffer, categoryName);
+                if (bytesWritten > 0)
+                {
+                    mappedTableName = stackBuffer.Slice(0, bytesWritten).ToArray();
+
+                    tableNameCache.CachedSanitizedTableNameCount++;
+                }
+                else
+                {
+                    // Note: When the table name could not be sanitized we cache
+                    // the empty array NOT s_passthroughTableName.
+                    mappedTableName = Array.Empty<byte>();
+                }
+            }
+
+            // Note: This is using copy-on-write pattern to keep the happy
```

Comment:
👍

Reply:
Mentioned above possibly using `Hashtable`. [Hashtable has a unique quirk in that it supports the many-readers-single-writer pattern](https://learn.microsoft.com/dotnet/api/system.collections.hashtable#thread-safety). That would allow removing copy-on-write, which is great for the warm-up phase:

|      Method |       Mean |    Error |   StdDev |      Gen0 |     Gen1 |   Allocated |
|------------ |-----------:|---------:|---------:|----------:|---------:|------------:|
| CopyOnWrite | 3,659.8 us | 63.39 us | 56.19 us | 1328.1250 | 183.5938 | 16326.84 KB |
|   Hashtable |   103.6 us |  1.95 us |  1.82 us |   13.4277 |   6.5918 |    165.3 KB |

<details>
<summary>Benchmark code</summary>

```csharp
using System.Collections;
using System.Collections.Generic;
using BenchmarkDotNet.Attributes;

namespace OpenTelemetry.Exporter.Geneva.Benchmark;

[MemoryDiagnoser]
public class CopyOnWriteBenchmarks
{
    private readonly object lockObject = new();

    [Benchmark]
    public Dictionary<string, byte[]> CopyOnWrite()
    {
        Dictionary<string, byte[]> d = new();

        for (int i = 0; i < 1024; i++)
        {
            string k = $"Category{i}";

            if (!d.TryGetValue(k, out var value))
            {
                lock (this.lockObject)
                {
                    if (!d.TryGetValue(k, out value))
                    {
                        d[k] = new byte[1];

                        var old = d;

                        d = new(old);
                    }
                }
            }
        }

        return d;
    }

    [Benchmark]
    public Hashtable Hashtable()
    {
        Hashtable ht = new();

        for (int i = 0; i < 1024; i++)
        {
            string k = $"Category{i}";

            var value = (string)ht[k];
            if (value == null)
            {
                lock (this.lockObject)
                {
                    value = (string)ht[k];
                    if (value == null)
                    {
                        ht[k] = new byte[1];
                    }
                }
            }
        }

        return ht;
    }
}
```
</details>

Chatted with @utpilla we decided to not use `Hashtable` for now because the docs also say it isn't recommended to be used.

It is an implementation detail so we can always change it in the future.